Add dropdown classes for contextual background colors

tordans opened the pull request:
The [Dropdown](http://v4-alpha.getbootstrap.com/components/button-dropdown/) Element should have the ability to receive colored backgrounds similar to the [Table contextual classes](http://v4-alpha.getbootstrap.com/content/tables/#contextual-classes) or [card background variants](http://v4-alpha.getbootstrap.com/components/card/#background-variants).
## Example Code

See `docs/components/dropdowns.md`
## Changes
- There are three special cases I see for the background colors, that make the whole change more complex that I like.
  - Edge case 1: colored item as first/last child; this the easiest and cleanest to solve.
  - Edge case 2: a divider, then a colored item; this is annoying, since there is now "previous sibling" in css yet, so I worked around it with negative margins and border. To keep the margins in sync they are extracted in a variable.
  - Edge case 3: a colored item, then a divider; the is also annoying, since the divider is not build in a way that would make this color change easy, so I worked around it with a before element. A cleaner way might be to find a way to refactor the divider to behave like three stacked color bars on top of each other; maybe by just switching the spacing from margins to borders?
- To solve edge case 3 I change the `.dropdown-divider`-overflow and position behavior. Also I removed/re-integrated `mixins/nav-divider` since it was only used by the divider anyways.
- The colors for the dropdown modifier are added to the variables and re-use the form-state-colors as a default.
- The dropdown documentation is extend i a very basic way.
## Questions
- Does the whole thing compile? I don't have a full setup here. // See comments below.
- Should it also work for the whole `.dropdown-menu` like it does for table>tr as well as table>td?
- Should the docs for [Buttons dropdown](http://v4-alpha.getbootstrap.com/components/button-dropdown/) and [Dropdown](http://v4-alpha.getbootstrap.com/components/dropdowns/) interlink more? Right now they duplicate information but there are no links for the information that is missing in the other place.
- Should there be an .inverted case?
- For the docs: What is the proper pattern to name the headlines that handle color changes. For tables, cards, … they all use different headlines like ["Contextual variations"](http://v4-alpha.getbootstrap.com/components/label/#contextual-variations), ["Examples"](http://v4-alpha.getbootstrap.com/components/buttons/#examples), ["Contextual classes"](http://v4-alpha.getbootstrap.com/content/tables/#contextual-classes) and so on. Maybe I should extract this in a separate issue?
- How to handle the fact that https://github.com/twbs/bootstrap/pull/13797 introduces colored :hover states that will cause new edge cases for this idea. Worst case is, that the colors are very similar so the :hover effect is nullified.

tordans committed:
Remove mixins/nav-divider

The mixing was used in one place only which was a unnecessary
abstraction.

tordans committed:
Remove double "// //"

This is probably just a mistake and not some kind of pattern, right?

tordans committed:
Add dropdown-<variant> background colors

Enable `.dropdown-item`s to have a background color the same as tables
and cards.

- Refactor the divider.
  - The margin is now a variable that we re-use inside the new mixin.
  - The overflow and position needs changing to work around the
divider-background-color.

- Add dropdowns mixing for background color. For now it only supports
the `.dropdown-item`, not the whole dropdown.
  - Edge case 1: colored item as first/last child
  - Edge case 2: colored item with a divider before the colored item
  - Edge case 3: colored item with a divider after the colored item
  - Add variables for the colors which per default are re-used from the
form-state-colors.

- Add documentation on the dropdown docs.

cvrebert commented:
X-Ref: #13797

cvrebert commented:
> Does the whole thing compile?

No, there's a compile error. See the Travis Details link (i.e. https://travis-ci.org/twbs/bootstrap/jobs/125335121 ).

> What is the proper pattern to name the headlines that handle color changes.

"Contextual variations" or "Contextual classes". We're not currently consistent as to which of those two.

> Should the docs for Buttons dropdown and Dropdown interlink more?

See #18545.

cvrebert commented:
CC: @mdo

tordans committed:
Fix compiling error due to wrong variable name

https://travis-ci.org/twbs/bootstrap/jobs/125335121

tordans commented:
@cvrebert I fixed the first travis error, the current one is due caused by the order of variables. Is it OK to move the whole "Form states and alerts" up a few lines? Or should I duplicate the colors for this PR?

cvrebert commented on scss/_variables.scss:
```diff
@@ -435,9 +435,15 @@ $dropdown-min-width:             160px !default;
 $dropdown-bg:                    #fff !default;
+$dropdown-primary-bg:            lighten($brand-primary, 60%); // todo, find a good color or remove the primary-bg
+$dropdown-success-bg:            $state-success-bg;
+$dropdown-info-bg:               $state-warning-bg;
```
info ≠ warning

cvrebert commented on scss/mixins/_dropdowns.scss:
```diff
@@ -0,0 +1,31 @@
+    background-color: $background;
+  }
+  // the divider that comes after a dropdown-item with color
+  .dropdown-#{$state} + .dropdown-divider:before {
```
Use `::before` instead of `:before`

cvrebert commented on scss/mixins/_dropdowns.scss:
```diff
@@ -0,0 +1,31 @@
+    margin-top: -$dropdown-divider-margin-y;
+    border-top: $dropdown-divider-margin-y $background solid;
+  }
+  .dropdown-item.dropdown-#{$state}:first-child {
```
Remove the `.dropdown-item` overqualification

cvrebert commented:
@tordans For now, I'd use the `$brand-*` variables instead of the `$state-*` variables.

tordans committed:
Fix and improve some more details

- cvrebert: „Use ::before instead of :before“
- cvrebert: „Remove the .dropdown-item overqualification“
- cvrebert: „info ≠ warning“
- cvrebert: „For now, I'd use the $brand-* variables instead of the
$state-* variables.“; added todo to add nicer colors later

tordans committed:
Fix css lint error: Property order

`scss/_dropdown.scss:61 [W] PropertySortOrder: Properties should be
ordered position, height, margin, background-color`
https://travis-ci.org/twbs/bootstrap/jobs/125341013

tordans commented:
@cvrebert thanks for the review, I change the code accordingly.

vsn4ik commented on scss/_variables.scss:
```diff
@@ -436,10 +436,10 @@ $dropdown-padding-y:             5px !default;
-$dropdown-warning-bg:            $state-warning-bg;
-$dropdown-danger-bg:             $state-danger-bg;
+$dropdown-success-bg:            lighten($brand-success, 60%); // todo, use $state-success-bg or something similar but those are specified below
+$dropdown-info-bg:               lighten($brand-info, 60%);    // todo, use $state-info-bg    or something similar but those are specified below
```
A lot of white spaces after `$state-info-bg` in comment.

vsn4ik commented on scss/_variables.scss:
```diff
@@ -436,10 +436,10 @@ $dropdown-padding-y:             5px !default;
+$dropdown-success-bg:            lighten($brand-success, 60%); // todo, use $state-success-bg or something similar but those are specified below
+$dropdown-info-bg:               lighten($brand-info, 60%);    // todo, use $state-info-bg    or something similar but those are specified below
+$dropdown-warning-bg:            lighten($brand-warning, 60%); // todo, use $state-warning-bg or something similar but those are specified below
+$dropdown-danger-bg:             lighten($brand-danger, 60%);  // todo, use $state-danger-bg  or something similar but those are specified below
```
A lot of white spaces after `$state-danger-bg` in comment.

mdo commented:
Going to punt on this, thanks though.

We've been asked to add this in the past I believe, but I don't see any real utility in multiple options for background colors on dropdowns. I understand the need or want to change the background for *all* dropdown items, which you can do with existing variables, but not on a per-case basis.

tordans commented:
@mdo Thanks for you answer. Here are the use cases I see and have in my app:
- differentiate features in a dropdown that are visible for admins vs. all users
- group items in a dropdown that belong together semantically
- highlight new feature in a dropdown that we want to market as a special new thing
- provide the setup to add a different hover-color to a dropdown item, like a delete entry

Are those all irrelevant for you?
